src/Compilers/CSharp/Test/Emit/Emit/DeterministicTests.cs

```diff
@@ -184,6 +185,54 @@ public void TestWriteOnlyStream()
             compilation.Emit(output);
         }
 
+        [Fact, WorkItem(11990, "https://github.com/dotnet/roslyn/issues/11990")]
+        public void ForwardedTypesAreEmmittedInADeterministicOrder()
+        {
+            const int generatedTypes = 100;
+
+            var forwardedToCode = new StringBuilder();
+            forwardedToCode.AppendLine("namespace ForwardedToNamespace {");
+            for (var i = 1; i <= generatedTypes; i++)
+            {
+                forwardedToCode.AppendLine($"public class T{i} {{}}");
+            }
+            forwardedToCode.AppendLine("}");
+
+            var forwardedToCompilation = CreateCompilation(forwardedToCode.ToString());
+            var forwardedToReference = new CSharpCompilationReference(forwardedToCompilation);
+
+            IEnumerable<string> compileAndGetExportedTypes()
+            {
+                var forwardingCode = new StringBuilder();
+                forwardingCode.AppendLine("using System.Runtime.CompilerServices;");
+                for (var i = 1; i <= generatedTypes; i++)
+                {
+                    forwardingCode.AppendLine($"[assembly: TypeForwardedTo(typeof(ForwardedToNamespace.T{i}))]");
+                }
+
+                var forwardingCompilation = CreateCompilationWithMscorlib(forwardingCode.ToString(), new MetadataReference[] { forwardedToReference });
+
+                using (var stream = forwardingCompilation.EmitToStream())
+                {
+                    using (var block = ModuleMetadata.CreateFromStream(stream))
+                    {
+                        foreach (var handle in block.MetadataReader.ExportedTypes)
+                        {
+                            var type = block.MetadataReader.GetExportedType(handle);
+                            var typeName = block.MetadataReader.GetString(type.Name);
+                            yield return typeName;
+                        }
+                    }
+                }
+            }
+
+            var baseline = compileAndGetExportedTypes().ToArray();
+            Assert.Equal(generatedTypes, baseline.Length);
+
+            var reference = compileAndGetExportedTypes().ToArray();
+            Assert.Equal(baseline, reference);
```

Comment:
@AlekseyTs The test doesn't verify they're sorted. The test verifies they are deterministic (two runs match). I'll add another test that has a fewer types (not a hundred) unsorted, and verifies they are sorted (in a single run).